crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1261,18 +1261,48 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
         event_id: Option<&EventId>,
         transaction_id: Option<&TransactionId>,
     ) -> Option<Arc<TimelineItem>> {
-        // Start with the canonical case: detect a local timeline item that matches
-        // `event_id` or `transaction_id`.
-        if let Some((local_timeline_item_index, local_timeline_item)) =
-            rfind_event_item(items, |event_timeline_item| {
-                if event_timeline_item.is_local_echo() {
-                    event_id == event_timeline_item.event_id()
-                        || (transaction_id.is_some()
-                            && transaction_id == event_timeline_item.transaction_id())
+        // Detect a local timeline item that matches `event_id` or `transaction_id`.
+        if let Some((local_timeline_item_index, local_timeline_item)) = items
+            .iter()
+            // Get the index of each item.
+            .enumerate()
+            // Iterate from the end to the start.
+            .rev()
+            // Use a `Iterator::try_fold` to produce a single value, and to stop the iterator
+            // when a non local event timeline item is met. We want to stop iterating when:
+            //
+            // - a duplicate local event timeline item has been found,
+            // - a non local event timeline item is met,
+            // - a non event timeline is met.
+            //
+            // Indeed, it is a waste of time to iterate over all items in `items`. Local event
+            // timeline items are necessarily at the end of `items`: as soon as they have been
+            // iterated, we can stop the entire iteration.
+            .try_fold((), |(), (nth, timeline_item)| {
```

Comment:
I think you can use a `find_map()` and not have to bump the minimum rust version, and it would be a bit more noob-friendly :)

Reply:
[`Iterator::find_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find_map) is not appropriate here. It returns the first non-`None` value. It means that we cannot differentiate the _break the loop because we've found a value_, vs. _break the loop because no value can be found_.

We need to stop the iteration when all local event timeline items have been traversed. This is not possible to represent that with `Iterator::find_map`. That's exactly the problem this patch is trying to fix :-).

Reply:
Why not a plain loop then… :innocent:

Reply:
That, that's a good question ;-).